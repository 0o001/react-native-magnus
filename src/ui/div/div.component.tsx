import * as React from "react";
import { useContext } from "react";
import {
  View as RNView,
  ImageBackground as RNImageBackground
} from "react-native";

import { DivProps } from "./div.type";
import { getStyle } from "./div.style";
import { ThemeContext } from "../../theme";

const Div: React.FunctionComponent<DivProps> = (props: DivProps) => {
  const {
    bg,
    h,
    w,
    m,
    mt,
    mr,
    mb,
    ml,
    p,
    pr,
    pt,
    pb,
    pl,
    minH,
    minW,
    maxW,
    maxH,
    position,
    style,
    flexDir,
    row,
    rounded,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    children,
    bgImg,
    bgMode,
    alignItems,
    justifyContent,
    borderColor,
    borderBottomColor,
    borderLeftColor,
    borderTopColor,
    borderRightColor,
    borderWidth,
    borderLeftWidth,
    borderRightWidth,
    borderBottomWidth,
    borderTopWidth,
    flexWrap,
    shadow,
    shadowColor,
    opacity,
    overflow,
    top,
    left,
    right,
    bottom,
    zIndex,
    ...rest
  } = props;
  const theme = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);

  if (bgImg) {
    return (
      <RNImageBackground
        source={bgImg}
        style={computedStyle.div}
        resizeMode={props.bgMode}
        imageStyle={computedStyle.image}
        {...rest}
      >
        {children}
      </RNImageBackground>
    );
  }

  return (
    <RNView style={computedStyle.div} {...rest}>
      {children}
    </RNView>
  );
};

Div.defaultProps = {
  bg: "transparent",
  flexDir: "column",
  flexWrap: "nowrap",
  rounded: "none",
  shadow: 0,
  shadowColor: "gray900",
  position: "relative",
  bgMode: "cover",
  pointerEvents: "auto",
  row: false,
  borderStyle: "solid"
};

export { Div };
