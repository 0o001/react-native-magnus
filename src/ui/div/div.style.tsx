import { StyleSheet } from "react-native";

import {
  getThemeProperty,
  createFlexStyles,
  createPositionStyle,
  createSpacingStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
  createBorderRadiusStyles
} from "../../theme/theme.service";

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.div = {
    flexDirection: props.flexDir,
    flexWrap: props.flexWrap,
    alignItems: props.alignItems,
    justifyContent: props.justifyContent,
    alignSelf: props.alignSelf,
    backgroundColor: getThemeProperty(theme.colors, props.bg, "transparent"),
    ...createFlexStyles(props),
    ...createPositionStyle(props),
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius)
  };

  computedStyle.image = {
    ...createBorderRadiusStyles(props, theme.borderRadius)
  };

  if (props.shadow) {
    computedStyle.div = {
      ...computedStyle.div,
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, "white")
    };
  }

  if (props.h) {
    computedStyle.div = {
      ...computedStyle.div,
      height: props.h
    };
  }

  if (props.w) {
    computedStyle.div = {
      ...computedStyle.div,
      width: props.w
    };
  }

  if (props.minH) {
    computedStyle.div = {
      ...computedStyle.div,
      minHeight: props.minH
    };
  }

  if (props.minW) {
    computedStyle.div = {
      ...computedStyle.div,
      minWidth: props.minW
    };
  }

  if (props.opacity) {
    computedStyle.div = {
      ...computedStyle.div,
      opacity: props.opacity
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.div = {
      ...computedStyle.div,
      ...props.style
    };
  }

  return StyleSheet.create(computedStyle);
};
