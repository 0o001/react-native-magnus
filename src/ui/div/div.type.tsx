import {
  ViewProps as RNViewProps,
  ImageSourcePropType as RNImageSourcePropType,
} from 'react-native';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
  ShadowPropsType,
  ResponsiveType,
  BaseBgModeType,
  BaseJustifyType,
  BaseAlignSelfType,
  BaseAlignItemsType,
  BaseFlexWrapType,
  BaseFlexDirectionType,
  BasePositionType,
  BaseOverflowType,
  BaseBorderStyleType,
} from '../../theme';

export interface DivProps
  extends RNViewProps,
    BorderPropsType,
    SpacingPropsType,
    RoundedPropsType,
    ShadowPropsType {
  h?: ResponsiveType<string | number>;
  w?: ResponsiveType<string | number>;
  bg?: ResponsiveType<string>;
  minH?: ResponsiveType<string | number>;
  minW?: ResponsiveType<string | number>;
  maxH?: ResponsiveType<string | number>;
  maxW?: ResponsiveType<string | number>;
  bgImg?: RNImageSourcePropType;
  bgMode?: ResponsiveType<BaseBgModeType>;
  flex?: ResponsiveType<number>;
  justifyContent?: ResponsiveType<BaseJustifyType>;
  alignSelf?: ResponsiveType<BaseAlignSelfType>;
  alignItems?: ResponsiveType<BaseAlignItemsType>;
  flexDir?: ResponsiveType<BaseFlexDirectionType>;
  flexWrap?: ResponsiveType<BaseFlexWrapType>;
  position?: ResponsiveType<BasePositionType>;
  overflow?: ResponsiveType<BaseOverflowType>;
  borderStyle?: ResponsiveType<BaseBorderStyleType>;
  top?: ResponsiveType<number>;
  right?: ResponsiveType<number>;
  bottom?: ResponsiveType<number>;
  left?: ResponsiveType<number>;
  opacity?: ResponsiveType<number>;
  children?: React.ReactNode[] | React.ReactNode;
  row?: ResponsiveType<boolean>;
  zIndex?: ResponsiveType<number>;
}
