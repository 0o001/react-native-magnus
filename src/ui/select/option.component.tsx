import * as React from "react";
import { TouchableHighlightProps as RNButtonProps } from "react-native";

import { Button } from "../button/button.component";

interface TextInputProps extends RNButtonProps {
  m?: any;
  p?: any;
  center?: boolean;
  value: any;
  onSelect?: (value: any) => void;
  children?: React.ReactElement[] | React.ReactElement;
}

type Ref = {};

const Option = React.forwardRef<Ref, TextInputProps>(props => {
  const { children, center, value, onSelect, ...rest } = props;

  /**
   * on press select
   *
   * @param e
   */
  const onPress = () => {
    if (onSelect) {
      onSelect(value);
    }
  };

  return (
    <Button {...rest} onPress={onPress} center={center} block>
      {children}
    </Button>
  );
});

Option.defaultProps = {
  onSelect: () => {},
  center: false
};

export { Option };
