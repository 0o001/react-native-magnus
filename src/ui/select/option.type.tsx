import { PressableProps as RNButtonProps } from 'react-native';
import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
  ShadowPropsType,
} from '../../theme';

export interface OptionProps
  extends RNButtonProps,
    BorderPropsType,
    SpacingPropsType,
    ShadowPropsType,
    RoundedPropsType {
  h?: number | string;
  w?: number | string;
  bg?: string;
  position?: 'absolute' | 'relative';
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline';
  flexDir?: 'row' | 'column' | 'row-reverse' | 'column-reverse';
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  top?: number;
  flex?: number;
  left?: number;
  right?: number;
  bottom?: number;
  color?: string;
  loading?: boolean;
  disabled?: boolean;
  loaderColor?: string;
  minW?: number | string;
  minH?: number | string;
  fontSize?: string | number;
  loaderSize?: number | string;
  suffix?: React.ReactNode;
  block?: boolean;
  borderless?: boolean;
  rippleColor?: string;
  center?: boolean;
  ripple?: boolean;
  value: any;
  onSelect?: (value: any) => void;
  selectedValue?: any;
}
