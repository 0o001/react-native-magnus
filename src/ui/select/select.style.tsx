import { StyleSheet } from 'react-native';

import {
  getThemeProperty,
  createSpacingStyles,
  createBorderRadiusStyles,
  createBorderColorStyles,
  createBorderWidthStyles,
} from '../../theme/theme.service';
import { ScreenHeight } from '../../utilities/dimension';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.wrapper = {
    backgroundColor: getThemeProperty(theme.colors, props.bg),
    ...createBorderWidthStyles(props),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius),
    height: ScreenHeight * 0.7,
  };

  computedStyle.indicator = {
    alignSelf: 'center',
    marginVertical: 10,
  };

  computedStyle.container = {
    flex: 1,
    ...createSpacingStyles(props, theme.spacing),
  };

  if (props.h) {
    computedStyle.container = {
      ...computedStyle.container,
      height: props.h,
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.container = {
      ...computedStyle.container,
      ...props.style,
    };
  }
  return StyleSheet.create(computedStyle);
};
