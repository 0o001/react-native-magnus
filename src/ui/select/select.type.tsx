import {
  ViewProps as RNViewProps,
  ImageSourcePropType as RNImageSourcePropType,
} from 'react-native';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from '../../theme';

export interface SelectProps
  extends RNViewProps,
    BorderPropsType,
    SpacingPropsType,
    RoundedPropsType {
  h?: number | string;
  title?: string | React.ReactNode;
  message?: string | React.ReactNode;
  w?: number | string;
  bg?: string;
  minH?: number;
  minW?: number;
  bgImg?: RNImageSourcePropType;
  showScrollIndicator?: boolean;
  bgMode?: 'contain' | 'cover' | 'stretch';
  flex?: number;
  shadow?: number;
  shadowColor?: string;
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline';
  flexDir?: 'row' | 'column' | 'row-reverse' | 'column-reverse';
  flexWrap?: 'wrap' | 'nowrap' | 'wrap-reverse';
  position?: 'absolute' | 'relative';
  top?: number;
  right?: number;
  bottom?: number;
  left?: number;
  multiple: boolean;
  value: any;
  children: React.ReactElement[] | React.ReactElement;
  footer: React.ReactElement[] | React.ReactElement;
  onSelect: (value: any) => void;
  data: any[];
  renderItem: (item: any, index: number) => React.ReactElement;
}
