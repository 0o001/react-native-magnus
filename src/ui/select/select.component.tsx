import * as React from 'react';
import RNModal from 'react-native-modal';
import { SafeAreaView, FlatList } from 'react-native';
import { useContext, useState, useImperativeHandle, useEffect } from 'react';

import { getStyle } from './select.style';
import { Div } from '../div/div.component';
import { ThemeContext } from '../../theme';
import { Text } from '../text/text.component';
import { Input } from '../input/input.component';
import { Icon } from '../icon/icon.component';
import { Option } from './select.option.component';
import { Button } from '../button/button.component';
import { SelectProps, SelectRef, CompoundedSelect } from './select.type';

const Select = React.forwardRef<SelectRef, SelectProps>((props, ref) => {
  const {
    value,
    title,
    footer,
    data,
    multiple,
    renderItem,
    keyExtractor,
    searchableProps,
    onSelect: onSelectProp,
  } = props;
  const { theme } = useContext(ThemeContext);
  const [visible, setVisible] = useState(props.isVisible || false);
  const [selectedValue, setSelectedValue] = useState(value);
  const [searchTerm, setSearchTerm] = useState<string>('');

  const computedStyle = getStyle(theme, props);

  const resolveMultiLevelAccess = (obj: any, keys: string) => {
    return keys.split('.').reduce((cur: any, key: string) => {
      return cur[key];
    }, obj);
  };

  const filterData = (data: any[], searchTerm: string) => {
    if (!searchableProps || !searchableProps.length) return data;

    var lowSearch = searchTerm.toLowerCase();
    return data.filter((item) => {
      return searchableProps.some((key) =>
        String(resolveMultiLevelAccess(item, key))
          .toLowerCase()
          .includes(lowSearch)
      );
    });
  };

  const filteredData = React.useMemo(() => filterData(data, searchTerm), [
    data,
    searchTerm,
  ]);

  useEffect(() => {
    if ('isVisible' in props) {
      setVisible(props.isVisible || visible);
    }

    if (!visible) setSearchTerm('');
  }, [props, visible]);

  /**
   * exposing functions to parent
   */
  useImperativeHandle(ref, () => ({
    close() {
      setVisible(false);
    },
    open() {
      setVisible(true);
    },
  }));

  /**
   * set values of select based if its multiple select
   * or single valued select
   *
   * @param value
   */
  const onSelect = (value: any) => {
    let finalValue;

    if (multiple) {
      const copy = selectedValue.slice();
      const index = selectedValue.indexOf(value);
      if (index !== -1) {
        copy.splice(index, 1);
      } else {
        copy.push(value);
      }

      setSelectedValue(copy);
      finalValue = copy;
    } else {
      setSelectedValue(value);
      setVisible(false);
      finalValue = value;
    }

    onSelectProp(finalValue);
  };

  /**
   * render title at top select modal
   */
  const renderTitle = () => {
    if (title) {
      return typeof title === 'string' ? (
        <Text fontSize="md" color="gray700" px="xl" pt="md" pb="lg">
          {title}
        </Text>
      ) : (
        title
      );
    }

    return false;
  };

  /**
   * render searchbar at top select modal
   */
  const renderSearchbar = () => {
    if (searchableProps?.length) {
      return (
        <Input
          prefix={<Icon mr="lg" name="search" color="gray700" fontSize="3xl" />}
          suffix={
            <Button
              h={30}
              w={30}
              rounded="circle"
              bg="gray700"
              onPress={() => setSearchTerm('')}
            >
              <Icon name="clear" color="white" fontSize="xl" />
            </Button>
          }
          px="xl"
          py="md"
          mx="xl"
          mt="md"
          mb="xl"
          pr="lg"
          value={searchTerm}
          onChangeText={(text) => setSearchTerm(text)}
          fontSize="md"
          borderWidth={0}
          placeholder="Search items"
          bg="gray200"
        />
      );
    }

    return false;
  };

  const renderFooter = () => {
    if (footer) return footer;

    // if the component is single-valued and no footer is provided
    // don't render anything in footer
    if (!multiple) return false;

    return (
      <Button
        block
        bg="green600"
        mx="xl"
        mt="lg"
        onPress={() => {
          setVisible(false);
        }}
      >
        Submit
      </Button>
    );
  };

  return (
    <RNModal
      backdropTransitionOutTiming={0}
      animationIn="slideInUp"
      isVisible={visible}
      backdropColor="black"
      onBackdropPress={() => setVisible(false)}
      onBackButtonPress={() => setVisible(false)}
      hideModalContentWhileAnimating
      style={{
        margin: 0,
        justifyContent: 'flex-end',
      }}
    >
      <Div style={computedStyle.wrapper}>
        <SafeAreaView style={computedStyle.container}>
          <Div>
            {renderTitle()}
            {renderSearchbar()}
          </Div>

          {filteredData.length ? (
            <FlatList
              data={filteredData}
              keyExtractor={keyExtractor}
              renderItem={({ item, index }) =>
                React.cloneElement(renderItem(item, index), {
                  onSelect,
                  selectedValue,
                })
              }
            />
          ) : (
            <Div flex={1} pl="xl">
              <Text fontSize="lg">No results found.</Text>
            </Div>
          )}

          {renderFooter()}
        </SafeAreaView>
      </Div>
    </RNModal>
  );
}) as CompoundedSelect;

Select.defaultProps = {
  bg: 'white',
  rounded: 'none',
  flexDir: 'column',
  isVisible: false,
  // @ts-ignore
  keyExtractor: (item, index) => `${index}`,
};

Select.Option = Option;

export { Select };
