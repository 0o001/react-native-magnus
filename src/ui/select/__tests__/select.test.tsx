import React from "react";
import renderer from "react-test-renderer";

import { Div } from "../../div/div.component";
import { Select } from "../select.component";
import { ThemeProvider } from "../../../theme";

describe("Div Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Select
          value={1}
          onSelect={() => {}}
          data={[]}
          renderItem={() => <Div />}
        >
          <Div />
        </Select>
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();
  });
});
