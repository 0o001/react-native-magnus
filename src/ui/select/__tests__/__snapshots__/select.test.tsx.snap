// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Div Component should render without issues 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={true}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  visible={false}
>
  <View
    accessible={true}
    clickable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    hideModalContentWhileAnimating={true}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    pointerEvents="box-none"
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-end",
        "margin": 0,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 0,
          "flexDirection": "column",
          "flexWrap": "nowrap",
          "justifyContent": undefined,
          "maxHeight": 933.8,
          "position": "relative",
        }
      }
    >
      <RCTSafeAreaView
        emulateUnlessSupported={true}
        style={Object {}}
      >
        <RCTScrollView
          data={Array []}
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View />
        </RCTScrollView>
        <View
          accessible={true}
          clickable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "#50c620",
              "borderRadius": 6,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginHorizontal": 24,
              "marginTop": 12,
              "overflow": "hidden",
              "padding": 8,
              "position": "relative",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "flexDirection": "row",
                "justifyContent": "center",
                "position": "relative",
                "width": "100%",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 15,
                  "lineHeight": 18,
                  "textAlign": "right",
                }
              }
            >
              Submit
            </Text>
          </View>
        </View>
      </RCTSafeAreaView>
    </View>
  </View>
</Modal>
`;
